README.md

```diff
@@ -1,277 +1,1037 @@
+in advance. But what if there was a token that they had never heard of? This opportunity would be
+missed. With offer and ask based queries, these unheard of tokens would be returned by the first
+query and the arbitrage algorithms can decide which other assets to query based on those results. In
+short, offer and ask based queries allow for organic discovery of arbitrage composition paths.
+
+### Single Swap Addresses
+
+This dramatically improves the usability of the DEX since users/frontends don't need to manage a
+possibly infinite set of addresses for users.
+
+### Democratic Upgradability
+
+Upgrades to Cardano-Swaps can propagate through the ecosystem of users in a similarly democratic
+fashion as SPOs upgrading their pools to a new version of `cardano-node`. Since users can close
+their swaps at any time, whenever there is a potential upgrade, users can choose to close their
+current swaps and recreate them with the new contracts. There is never a bifurcation of liquidity
+due to the composable nature of all swaps, regardless of the protocol's version.
+
+### Frontend Agnosticism
+
+Thanks to the query-ability of beacon tokens, it is trivial for any frontend to integrate with
+Cardano-Swaps. For example, any wallet can integrate `cardano-swaps` by adding support for querying
+the beacon tokens. They can also add their own user friendly way to create and use swaps. The only
+requirement is that all frontends/users agree to use the same beacon token standard. There is no
+need for dedicated (censorable) frontends.
+
+### Naturally Resistant to Denial-of-Service Attacks
+
+Recall that each swap requires a deposit of at least 2 ada. While this behavior is due to the
+current protocol parameters, it is actually a good thing since it helps prevents denial-of-service
+attacks. For example, since the protocol does not check that the assets in the datum are real
+assets, it is possible to create swaps for fake assets. *This does not impact normal users who are
+querying swaps for a known trading pair.* The user would have to deliberately query a trading pair
+that contains a fake asset in order to see these fake UTxOs.
+
+These fake asset UTxOs only really impact offer and ask queries which are mainly for arbitragers who
+will likely have powerful hardware. Since these fake UTxOs still require 2 ada, creating millions of
+them (enough to disrupt the arbitragers) would require a deposit of millions of ada throughout the
+life of the denial-of-service attack. The only way to reclaim the ada is to stop the attack.
+Furthermore, checking if an asset is real is trivial (look for mint history). Once an arbitrager
+realizes an asset is fake, they can ignore it. So an actual denial-of-service attack would require
+constantly changing the millions of fake swaps to new fake swaps; this has a large cost in terms of
+transaction fees which makes this impractical to do at scale.
+
+
+## FAQ
```

Comment:
I'd add that in the endgame, _most_ users will not be "competing" for swaps - market makers / businesses will be. Most users will create orders via swap UTxOs that are then swept by larger players (MMs/arbitragers) - in this scenario there is zero concurrency problem.

Reply:
I respectfully would recommend that you simplify the explanation somehow because it seems like the description is making things sound worse than they really are.  The way I see things is that market makers are going to be searching for all opportunities to profit and therefore they will combine all the overlapping djed swaps that can be executed profitably.  I think the market makers will end up competing for essentially the same "meshwork" of profitable swaps they want to combine.  This is a slight irritation for the market makers who will see some transactions fail, particularly highly profitable ones that will get competed over.  But, it is not a problem for ordinary users who will quickly get their swaps executed if "in the money".

Reply:
How about this wording?

> ### If there is one two-way swap UTxO with 1000 djed and there are 10 users trying to use 100 djed each, isn't this still a concurrency issue despite having enough liquidity?                                     
>                                                                                                                                                                                                             
> Yes, this is a concurrency bottleneck but no, this is not an issue. This is the perfect niche for businesses to build on top of Cardano-Swaps. Historically, whenever there is a risk that users don't want to carry, a business can form to carry that risk instead, for a price. Insurance is one such example. The same scenario applies here.                                                                                                                                                                           
>                                                                                                                                                                                                                   
> Imagine a slightly simpler example: Alice has a 100 djed swap available while Mike and Charlie both want 50 djed from it. No liquidity bottleneck here, just a concurrency bottleneck. Sarah can see  this scenario and decide to front the 50 djed to both Mike and Charlie (possibly for a fee) with the intention of settling up with Alice's swap later, ideally for a profit. At a high-level, this is how market making works. Mike and Charlie will likely prefer going through Sarah, even if Sarah charged a fee, simply because it means they don't have the concurrency risk. *The option to directly interact with Alice's swap will always be available to Mike and Charlie.* Because this option is always available, Sarah is incentivized to give Mike and Charlie a fair deal. Tiered fees will make it even easier for Mike or Charlie to use Alice's swap directly, if needed.                                                                                                                                          
>                                                                                                                                                                                                                   
> Thanks to Cardano-Swaps being fully p2p and permissionless, anyone can build a business on it like Sarah. And since the day-to-day users will likely not want the concurrency risk, there is every incentive for people like Sarah to step in. In the endgame, *most* users will not be competing for swaps - market makers and other businesses will be. Most sellers will create orders via swap UTxOs that are then swept by larger players (MMs/arbitragers) while most buyers will go through businesses that carry the concurrency risk in their stead. In this scenario, there is zero concurrency problem for the end user.                                                                                                                                                                                                  
>                                                                                                                                                                                                                   
> When it comes to creating a decentralized economy, not every feature needs to be directly enabled by the DApp. Some features can, and should, be left to the market to fill.

Reply:
Both work equally well for me but I think the latter (shorter, more familiar) edit will be more approachable by general readers... even docs for technical audiences in crypto & other Internet apps still have the Alice & Bob stories 😄

Reply:
In my head, I see things playing out like this:

User Alice has a 100 djed swap available.  User Mike and user Charlie both want 50 djed from it, so they submit their swaps.  A market maker sees these three swaps on-chain and notices that the exchange price that both Mike and Charlie have set overlaps with the price that Alice wants.  Consequently the market maker submits a transaction that uses Alice's swap combined with both Mike's and Charlie's so that all three swaps are executed in one transaction and the market maker profits from the amount of price overlap minus the transaction fees.

Thus there isn't a concurrency issue because market makers will look for such trades and try to bundle as many as possible that overlap into one transaction whenever a profit can be made.  The only concurrency issue results from the market makers competing to be first to execute the entire linked bundle of profitable (overlapping price) trades.

Furthermore, since anyone can be a searcher or market maker, anyone can search for, bundle up, and execute, profitable trades.  The more that are profitable which can be combined into one transaction the lower the proportional fees for the searcher.

Reply:
> User Alice has a 100 djed swap available. User Mike and user Charlie both want 50 djed from it, so they submit their swaps. A market maker sees these three swaps on-chain and notices that the exchange price that both Mike and Charlie have set overlaps with the price that Alice wants.

This only works if Charlie and Mike both deliberately incentivize someone to execute their swaps. To put it concretely, they can't use the exact same price as Alice. They need to deliberately create an arbitrage opportunity. To use the example, Mike and Charlie can only ask for 49 djed each so that 1 djed can be claimed by the arbitrager for each swap.

I think this is valid, though. I hadn't considered deliberately creating an arbitrage opportunity. I always thought of arbitrage opportunities as an accident. For volatile pairs, you *might* be able to use the same price but if the main concern is immediate conversion, the better the incentive for arbitragers the faster it will be converted. 

I think I like this approach more even though both are valid. With this one, Charlie and Mike set the fee they are willing to pay for immediate conversion. Meanwhile, in the previous example, Sarah sets the fee.

Nice observation!

Reply:
@TerminadaPool I changed it to use your method. How's this:

> ### If there is one two-way swap UTxO with 1000 djed and there are 10 users trying to use 100 djed each, isn't this still a concurrency issue despite having enough liquidity?                                     
>                                                                                                                                                                                                              
> Yes, this is a concurrency bottleneck but no, this is not an issue. This is the perfect niche for businesses to build on top of Cardano-Swaps. Historically, whenever there is a risk that users don't want to carry, a business can form to carry that risk instead, for a price. Insurance is one such example. The same scenario applies here.                                                                                                                                                                           
>                                                                                                                                                                                                              
> Imagine a slightly simpler example: Alice has a 100 djed swap available while Mike and Charlie both want 50 djed from it. No liquidity bottleneck here, just a concurrency bottleneck. If Mike and Charlie do not want to deal with the concurrency issues, they can pay arbitragers to execute the swaps for them. Here is how it would work:                                                                                                                                                                         
>                                                                                                                                                                                                                 
> - Both Mike and Charlie create one-way swaps with prices so that, if executed against Alice's swap, Mike and Charlie would each get 49 DJED and whoever executes their swaps gets to keep the other 1 DJED per swap. These swaps are effectively saying: "I'll pay you 1 DJED to execute this swap for me."                                                                                                                                                                                                               
> - Sarah, an arbitrager, can see all three swaps and executes them, keeping the 2 DJED arbitrage (1 DJED from each swap).                                                                                                                                                                                              
>                                                                                                                                                                                                                
> In other words, Mike and Charlie *deliberately create an arbitrage opportunity* so that arbitragers will want to execute their swaps. If Charlie really needs his swap executed ASAP, he can set his price so that he gets 48 DJED and the arbitrager gets 2 DJED for executing his swap. Basically, the more attractive the buyer makes the arbitrage opportunity, the more arbitragers will compete for it and the faster their swap will be converted. In this scenario, neither Mike nor Charlie experience a concurrency bottleneck. The arbitragers competing for their swaps are the ones experiencing the concurrency bottleneck, and they are paid accordingly for taking on this risk.                                                                                                                                     
>                                                                                                                                                                                                      
> In the endgame, *most* users will not be competing for swaps - arbitragers and other businesses will be. Most users will create orders via swap UTxOs that are then swept by these businesses. In this scenario, there is zero concurrency problem for the end user.                                                                                                                                                      
>                                                                                                                                                                                                          
> When it comes to creating a decentralized economy, not every feature needs to be directly enabled by the DApp. Some features can, and should, be left to the market to fill.

Reply:
This is now the second question in the FAQ. I added a short paragraph noting the similarities with setting a limit order below the market price in TradFi. I personally did not notice this connection until @TerminadaPool brought it up so I don't want to assume others will immediately see the connection.

Reply:
I still think you are underselling Cardano-Swaps.  I love your project, so this is why I am trying to figure out how you can word this FAQ without underselling it.

How does this sound?:

Imagine a slightly simpler example: Alice has a 100 DJED -> 100 ADA swap available.  Mike and Charlie both want to swap 50 ADA for DJED.  If they post their swaps on-chain as follows 50 ADA -> 50 DJED then there is no profit motive for anyone to combine and execute these trades after including the small transaction fee.

Mike and Charlie can easily each create a new swap which consumes Alice's swap where they pay the transaction fee.  However, if they don't want to do the combining themselves they can easily create a swap with enough overlap in price so that an arbitrager will be incentivised to bundle up and execute their trades.  For example, Mike and Charlie could both create 49.8 ADA -> 50 DJED swaps.  This would enable an arbitrager to notice that there is a 0.2 ADA profit opportunity (less transaction fee).  The arbitrager can now combine all 3 swaps for a slight profit.  He only needs 98.6 ADA to get 100 DJED from completing both Mike's and Charlie's swaps and can use the 100 DJED he will receive to execute Alice's swap to receive 100 ADA.  This will net the arbitrager 0.4 ADA less approx 0.2 ADA transaction fee for a net profit of 0.2 ADA.  One thing to notice about this is that the more swaps the arbitrager can combine, the more profitable it can be since he will always only pay one transaction fee and this fee won't increase very much with each extra swap added.  Arbitragers will seek out and complete these swaps because even when the profit is small because there is no risk to them and it is free money.

Many users of Cardano-Swaps will simply list their swaps on-chain and wait for their set price to be reached, just like Alice in the example above who created a swap with an asking price of 1.00 DJED for ADA.  Listing swaps on-chain like this creates extra liquidity.  Arbitragers will constantly look for profitable trades and whenever they see a slight overlap they will know they can bundle all these overlapping swaps together for a small profit.

This is exactly how the order book works in traditional finance except that each user also has the option of being the market maker (arbitrager) themselves.  For example, Mike might list his 49.8 ADA -> 50 DJED swap on-chain and Charlie will be able to see this.  Charlie can choose to be a market maker and bundle up Mike's swap with his own requirement for 50 Ada.  Being a market maker will be easy enough that anyone can do it.  Charlie can take the 50 DJED from Mike's swap, contribute another 50 DJED of his own, and execute Alice's swap using the 100 DJED to receive 100 ADA of which 49.8 will go to completing Mike's swap.  Charlie now has 50.2 ADA less the transaction fee of 0.2 ADA for a net of 50 ADA.

You might think that this will create a lot of competition between arbitragers but it is likely that most users will simply look at the prevailing price of the intended swap and set their price accordingly.  If they want their trade executed quickly then they will set their price so that there is some profit motive for an arbitrager.  Professional arbitragers will run sophisticated searching algorithms looking for long chains of profitable swaps.  In some cases this will mean that the combined bundle of a long chain of swaps might be overall profitable even though some individual links in the chain might not be profitable in isolation.  In other words, it will be cheaper and easier for users to let more sophisticated arbitragers seek out and bundle up lots of swaps together.  Users can just set their price and let the arbitragers look for the cheapest way to combine their swap.

The good thing about this is that Alice, Mike and Charlie need only set the exchange price they are happy receiving and their swaps will get executed when they are 'in the money" after taking into account the transaction fee.  However, unlike the traditional finance world, they each preserve the option of being the market maker (arbitrager) themselves.

Reply:
>  He only needs 98.6 ADA

I know you mean 99.6 but the example still works very well 🤓 ... and I agree no reason to be advertising this arrangement as a limiting factor, because this is seen so extensively use today in
1. offering / asking a more fetching price for larger orders in an exchange crypto order book
2. offering a higher Bitcoin transaction fee to ensure a Tx is put into an early block

Reply:
Your examples seem too technical/mathematical. To make the FAQ as approachable as possible, I think examples should be simplified to the bare minimum.

> I still think you are underselling Cardano-Swaps.

I don't see much of a difference between your version and the current one. Which parts do you think the current version is underselling?

>  no reason to be advertising this arrangement as a limiting factor

Which parts make it come across as a limiting factor? I am personally not interpreting it as a limiting factor and, to me, the current version does not really read that way. I think this paragraph:

> Yes, this is a concurrency bottleneck but no, this is not an issue. This is the perfect niche for businesses to build on top of Cardano-Swaps. Historically, whenever there is a risk that users don't want to carry, a business can form to carry that risk instead, for a price. Insurance is one such example. The same scenario applies here.

can possibly be read in a negative light. I can remove it entirely (aside from the first sentence). Then this paragraph:

> Users can think of this approach as creating a limit order just below the market price so that it is immediately executed; the further the order is set below the market price, the faster it will be executed. Users can decide for themselves how much they are willing to pay for a quick execution and set their prices accordingly.

can be reworded to express how this practice is already very common like in these two examples:

> 1. offering / asking a more fetching price for larger orders in an exchange crypto order book
> 2.  offering a higher Bitcoin transaction fee to ensure a Tx is put into an early block

and I would also add tiered fees suggested for cardano.

Reply:
What do you think are the most important points this answer should convey?

To me, they are:
1. Users can pay to have businesses compete for UTxOs on their behalf. Users can set their own prices for this.
2. Users always have the option to interact with UTxOs directly if desired.
3. Anybody can be a liquidity provider and anybody can get paid to compete for UTxOs. It is fully permissionless. This competition also means users don't need to pay that much to have their swaps executed.

Reply:
Yes, sorry, 99.6.

I really think the reader needs to better understand just how good this design is in term of completely decentralising the role of arbitrage seeker.  Moreover, people need to understand that these arbitrage seekers will not be like the MEV bots on Ethereum that front-run users to steal from them.  Instead Cardano-Swaps makes it easy for anyone to be an arbitrage seeker and these bots will continually look for opportunities to bundle up sequences of swaps that result in a closed loop which is even minimally profitable.

And, why wouldn't lots of people run arbitrage seeker bots since even just a penny here or there is still risk free profit.  There is no disincentive to bundle up and execute a sequence of swaps that delivers even a tiny net fraction of some token after the transaction fee is paid.  The deterministic eUTxO model of Cardano means that the arbitrager can't lose.  Either the entire transaction goes through giving him his small fraction of extra tokens or it doesn't because someone else sniped the bundle first.  These arbitrage seekers are not "market makers" in the TradFi sense because they are not taking any inventory risk if they are simply closing out a circular chain of swaps.

So Cardano-Swaps gives users are choice:  They have the option to create a swap and then execute it themselves in the simplistic sense by paying the full transaction fee and bundling only the minimum swaps in order to fulfil their needs.  Or, they can simply submit their swap on-chain and let an arbitrager combine their swap with a number of others and thereby save some transaction fees depending on how aggressively they set their swap price.

I believe these arbitrage seekers are going to get very professional and some will even act as market makers in the TradFi sense because some will be prepared to hold an inventory of tokens and accept directional risk.  This will result in even cheaper costs for ordinary users because they will be able to set their swap prices more aggressively (ie. with less arbitrage profit).

Another difference with TradFi is that Cardano-Swaps makes all pricing, and all fees, transparent.  There is none of this selling of privileged access to the order book because everyone can see the order book on-chain.

Eventually, I can see how Cardano-Swaps could be combined with Hydra for increased throughput.  Imagine a stock exchange running a number of hydra heads with each broker having it's own hydra heads.  Clients can transfer their stock tokens to smart contracts managed by their broker's hydra heads.  The clients can trade on the stock exchange for minimal fees with each broker running bots that bundle up chains of Cardano-Swaps, even involving sequences of different trading pairs for enhanced liquidity.  When the client wants to cash out or stop trading for a while they can simply withdraw their tokens from the broker's Hydra head smart contracts into their own wallet for safe self custody.  This will absolutely crash the infrastructure costs of being a broker and remove the barriers to competition.  Furthermore, users will then demand the transparency guaranteed by these new brokers.  Since Hydra offers isomorphic compatibility with base layer Cardano, the Cardano-Swaps smart contracts can be used unchanged.

Reply:
> What do you think are the most important points this answer should convey?
> 
> To me, they are:
> 
> 1. Users can pay to have businesses compete for UTxOs on their behalf. Users can set their own prices for this.
> 2. Users always have the option to interact with UTxOs directly if desired.
> 3. Anybody can be a liquidity provider and anybody can get paid to compete for UTxOs. It is fully permissionless. This competition also means users don't need to pay that much to have their swaps executed.

Yes, I think so.

I also agree that you want to keep the answers as simple as possible.  It is your document, I am really just trying to provide a different perspective to stimulate your thought if that can help paint it in the best light.  All these other DEX designs are inferior and it seems that AXO will not be permissionless.  At the end of the day, I just want to see Cardano-Swaps go for number one!

Reply:
I definitely appreciate the perspectives so don't hesitate to keep sharing!

Your example has a slight issue:

> For example, Mike and Charlie could both create 49.8 ADA -> 50 DJED swaps.

If Alice's swap is 1:1, 49.8 ADA can only get you 49.8 DJED. Mike's and Charlie's swaps would actually have to be 50.2 ADA -> 50 DJED. So the arbitrager would have from the two swaps 100.4 ADA but only needs 100 ADA to get the 100 DJED to satisfy them. The extra 0.4 ADA is left for the arbitrager.

How's this wording using your example (I think your example makes it a little clearer):

> ### If there is one two-way swap UTxO with 1000 DJED and there are 10 users trying to use 100 DJED each, isn't this still a concurrency issue despite having enough liquidity?
>
> Yes, this is a concurrency bottleneck but no, this is not an issue. *Cardano-Swaps addresses these bottlenecks by natively supporting a healthy fee market that is immune to MEV and front-running.* 
>
> Imagine a slightly simpler example: Alice has a 100 DJED swap available for ADA at 1:1 and Mike and Charlie both want DJED from it. No liquidity bottleneck here, just a concurrency bottleneck. If Mike and Charlie do not want to deal with the concurrency bottleneck, they can pay arbitragers to execute the swaps for them. Here is how it would work:
> - Both Mike and Charlie create one-way swaps for 50.2 ADA -> 50 DJED (a price of about 0.99 DJED per ADA). If executed against Alice's swap, Mike and Charlie would each get 50 DJED and whoever executes their swaps gets to keep the other 0.2 ADA per swap. These swaps are effectively saying: "I'll pay you 0.2 ADA to execute this swap for me."                                                                                                                                                                          
> - Sarah, an arbitrager, can see all three swaps and executes them, keeping the 0.4 ADA arbitrage (0.2 ADA from Mike's swap and 0.2 ADA from Charlie's swap). After the 0.2 ADA transaction fee, she pockets the remaining 0.2 ADA. 
>
> In other words, Mike and Charlie deliberately create an arbitrage opportunity so that arbitragers will want to execute their swaps. If Charlie really needs his swap executed ASAP, he can configure his swap so that he gets 50 DJED and the arbitrager gets 0.5 ADA for executing his swap. Basically, the more attractive the buyer makes the arbitrage opportunity, the more arbitragers will compete for it and the faster their swap will be converted. 
>
> From the perspective of the arbitragers, the only cost to them is the transaction fee which scales very slowly relative to the number of swaps in the transaction (14 swaps costs about 1.5 ADA). Combine this with the fact that these transactions are risk free for arbitragers (a failed transaction is no cost to the arbitrager) and you end up with very healthy incentives for arbitragers to *want* to do this and satisfy as many users as possible with each transaction. The risk free nature means arbitragers would even be willing to do this for small profits like 0.2 ADA. Being an arbitrager is fully permissionless so anyone can start generating income this way. 
>
> These incentives are how Cardano-Swaps natively creates a fee market. It is effectively no different than paying a larger transaction fee to ensure your transaction gets into the next block. Furthermore, since users must explicitly set their prices which are trustlessly enforced, MEV and front-running are simply not possible. 
>
> As this scenario shows, most users will likely not be competing for swaps - arbitragers and other businesses will be. Most users will create orders via swap UTxOs that are then swept by these businesses. This results in there being zero concurrency problem for the end user. 
>
> When it comes to creating a decentralized economy, not every feature needs to be directly enabled by                                                                                                               
the DApp. Some features can, and should, be left to the market to fill.

I want the answer to be fairly short but still address the key points. To me the key take-away is that Cardano-Swaps has an endogenous fee market and that is how it solves the concurrency bottleneck. This fee market is immune to MEV and front-running (I think... I have not really thought this through yet.).

Reply:
Feel free to make up some more FAQs if you like.  For example, you could add an FAQ about using Cardano-Swaps with Hydra.

Reply:
I definitely want to mention hydra at some point but right now, with hydra not really finalized, there is too much speculation for that. Once I'm done with the protocols, I do want to try playing around with hydra and see what I can do with it. For now, I want to keep the FAQ to the criticisms I see most often. Since I also want to release it this weekend, it is a bit late to add much more...

Reply:
@fallen-icarus : That re-write above is awesome, it really showcases the strengths.  In the end, I think the arbitragers will be so actively looking for any profitable bundle of swaps that users will simply post swaps that are very minimally in the money if they need them quickly executed.  And, in reality any tiny profit margin is still risk free profit.  Even it is only a penny worth, as an arbitrager, you already have the software running so why not take that penny.